compiler/src/dotty/tools/dotc/semanticdb/TypeOps.scala

```diff
@@ -0,0 +1,305 @@
+          loop(const.typeValue)
+
+        case ConstantType(const) =>
+          s.ConstantType(const.toSemanticConst)
+
+        case rt @ RefinedType(parent, name, info) =>
+          // `X { def x: Int; def y: Int }`
+          // RefinedType(
+          //   parent = RefinedType(
+          //     parent = TypeRef(..., X)
+          //     ...
+          //   )
+          //   refinedName = x
+          //   refinedInfo = TypeRef(..., Int)
+          // )
+          type RefinedInfo = (core.Names.Name, Type)
+          def flatten(tpe: Type, acc: List[RefinedInfo]): (Type, List[RefinedInfo]) = tpe match {
+            case RefinedType(parent, name, info) =>
+              flatten(parent, acc :+ (name, info))
+            case _ =>
+              (tpe, acc)
+          }
+
+          // flatten parent types to list
+          // e.g. `X with Y with Z { refined }`
+          // RefinedType(parent = AndType(X, AndType(Y, Z)), ...)
+          // => List(X, Y, Z)
+          def flattenParent(parent: Type): List[s.Type] = parent match {
+            case AndType(tp1, tp2) =>
+              flattenParent(tp1) ++ flattenParent(tp2)
+            case _ => List(loop(parent))
+          }
+
+          val (parent, refinedInfos) = flatten(rt, List.empty)
+          val stpe = s.WithType(flattenParent(parent))
+
+          val decls = refinedInfos.flatMap { (name, _) =>
+            refinementSymtab.get((rt, name))
+          }
+          val sdecls = decls.sscope(using LinkMode.HardlinkChildren)
+          s.StructuralType(stpe, Some(sdecls))
+
+        case rec: RecType =>
+          loop(rec.parent) // should be handled as RefinedType
+
+        // repeated params: e.g. `Int*`, which is the syntax sugar of
+        // `Seq[Int] @Repeated` (or `Array[Int] @Repeated`)
+        // See: Desugar.scala and TypeApplications.scala
+        case AnnotatedType(AppliedType(_, targs), annot)
+          if (annot matches defn.RepeatedAnnot) && (targs.length == 1) =>
+          val stpe = loop(targs(0))
+          s.RepeatedType(stpe)
+
+        case ann: AnnotatedType if ann.annot.symbol.info.isInstanceOf[ClassInfo] =>
+          def flatten(tpe: Type, annots: List[Annotation]): (Type, List[Annotation]) = tpe match
+            case AnnotatedType(parent, annot) if annot.symbol.info.isInstanceOf[ClassInfo] =>
+              flatten(parent, annot +: annots)
+            case other => (other, annots)
+
+          val (parent, annots) = flatten(ann, List.empty)
+          val sparent = loop(parent)
+          val sannots = annots.map(a =>
+            s.Annotation(loop(a.symbol.info.asInstanceOf[ClassInfo].selfType))
+          )
+          s.AnnotatedType(sannots, sparent)
+
+        case app @ AppliedType(tycon, args) =>
+          val sargs = args.map(loop)
+          loop(tycon) match
+            case ref: s.TypeRef => ref.copy(typeArguments = sargs)
+            // is there any other cases?
```

Comment:
We can have curried type application with type lambdas so we need to support that, instead of just replacing the  previously applied type arguments

Reply:
You mean something like `def x[T](a: T): ([X] =>> List[X])[T] = ???` ?
For type lambdas, it looks in this phase, what we will have is just an already applied type: e.g. `List[T]` in the above example. There might be another type than `TypeRef` here, but I couldn't find the example...

Reply:
here is an example:
```scala
class HKClass[F <: [T] =>> [U] =>> (U, T)] {
  def foo[T,U](x: F[T][U]): String = x.toString()
}
```

Reply:
fixed https://github.com/lampepfl/dotty/pull/12885/commits/26fc7ca4165e433df468ccfda14c3d6325f9ce13